src/llvm-cpufeatures.cpp

```diff
@@ -0,0 +1,151 @@
+// This file is a part of Julia. License is MIT: https://julialang.org/license
+//
+// Lower intrinsics that expose subtarget information to the language. This makes it
+// possible to write code that changes behavior based on, e.g., the availability of
+// specific CPU features.
+//
+// The following intrinsics are supported:
+// - julia.cpu.have_fma.$typ: returns 1 if the platform supports hardware-accelerated FMA.
+//
+// Some of these intrinsics are overloaded, i.e., they are suffixed with a type name.
+// To extend support, make sure codegen (in intrinsics.cpp) knows how to emit them.
+//
+// XXX: can / do we want to make this a codegen pass to enable querying TargetPassConfig
+//      instead of using the global target machine?
+
+#include "llvm-version.h"
+
+#include <llvm/IR/Module.h>
+#include <llvm/IR/Constants.h>
+#include <llvm/IR/Instructions.h>
+#include <llvm/IR/PassManager.h>
+#include <llvm/IR/LegacyPassManager.h>
+#include <llvm/Target/TargetMachine.h>
+#include <llvm/Support/Debug.h>
+
+#include "julia.h"
+
+#define DEBUG_TYPE "cpufeatures"
+
+using namespace llvm;
+
+extern TargetMachine *jl_TargetMachine;
+
+// whether this platform unconditionally (i.e. without needing multiversioning) supports FMA
+Optional<bool> always_have_fma(Function &intr) {
+    auto intr_name = intr.getName();
+    auto typ = intr_name.substr(strlen("julia.cpu.have_fma."));
+
+#if defined(_OS_WINDOWS_)
+    // FMA on Windows is weirdly broken (#43088)
+    return false;
+#elif defined(_CPU_AARCH64_)
+    return typ == "f32" || typ == "f64";
+#else
+    (void)typ;
+    return {};
+#endif
+}
+
+bool have_fma(Function &intr, Function &caller) {
+    auto unconditional = always_have_fma(intr);
+    if (unconditional.hasValue())
+        return unconditional.getValue();
+
+    auto intr_name = intr.getName();
+    auto typ = intr_name.substr(strlen("julia.cpu.have_fma."));
+
+    Attribute FSAttr = caller.getFnAttribute("target-features");
+    StringRef FS =
+        FSAttr.isValid() ? FSAttr.getValueAsString() : jl_TargetMachine->getTargetFeatureString();
```

Comment:
The access to jl_TargetMachine here is problematic. It should instead use the TLI to query the TM for the features, so that this pass can be run through opt

Reply:
It wasn't clear to me how to do getAnalysisUsage(TargetLibraryInfoWrapperPass) with the new pass manager infrastructure here.

Reply:
I think the canonical way is something like https://github.com/JuliaLang/julia/pull/42463/files#diff-8256bffe96cd66c411bbca57c488532cdba63fc13d346548181f7a303dec5e9aR2689-R2695

https://github.com/llvm/llvm-project/blob/5a667c0e741e5a895161b7a14376c59632fc5aa1/llvm/lib/Target/AMDGPU/AMDGPUPrintfRuntimeBinding.cpp#L579-L602

Reply:
In conversation with decided that `TargetLibraryInfo` is actually the wrong thing, but `TargetTransformInfo` might be the right thing. It already has things like https://llvm.org/doxygen/classllvm_1_1TargetTransformInfo.html#ac31bf22f119c5a99c36646d8e0eb2c0f, but it doesn't expose information about FMA that we might have to add ourselves first.

Alternativly this pass could take the TM on construction, still meh.

Reply:
So, it turns out, in offline discussion, that this possibly wasn't even the correct variable to look at anyways. We instead probably wanted to query the processor.cpp file to ask whether the current running system image was compiled with support for `+fma` in the architecture flags. Though we may want to now make it incompatible to load a "-fma" image onto a "+fma" processor (now similar to having a vector-size mismatch)?

Reply:
This is a long way from my expertise, but would it make sense to add this as an LLVM intrinsic? We can't be the only ones who want something like this.

Reply:
> Though we may want to now make it incompatible to load a "-fma" image onto a "+fma" processor (now similar to having a vector-size mismatch)?

Why should that be incompatible? Loading a non-fma image onto a fma processor should not cause any problems AFAICT.

There are two possible semantics for this intrinsic,

1. it returns result for the features the code is compiled for
2. it returns the result for the features available at runtime (which might also be different from the host processor one if the user disables it on the command line)

For the former, it's not possible for different calls of this function to return different values for different parts of the code, but as long as the information is used more or less locally it should not cause much problem. It'll still guarantee that if it returns true, the feature will be available on the processor and also available for the code around that to actually use.
For the latter it's not particularly useful when it returns a different result from the former since it's not generally possible to make llvm emit code in part of the function that uses features unavailable to the function. In another word, if this is the semantics of the intrinsics, when `has_fma()` returns true, it's not guaranteed that fma will be used for the code around it.

Taking the use of this intrinsic when making cloning decision will help a little, but is not a guarantee for consistency since there's no guarantee that the matching fma target is provided. Adding fma target for each non-fma target will fix this, and make the two semantics exactly the same, but that's not a scalable solution especially when done globally. It's not helped by the fact that on x86 there are two fma flavors (so you need two fma target per non-fma target). And it's definitely not scalable, as in it'll blow up exponentially, when feature detection for other features are added. Locally adding this isn't going to help very much either though since it'll rely too much on the inlining decision.

GCC avoids this scaling problem by having different notation for enabling features per function and detecting them, and I think that's what we need to do as well with some tweaks. The intrinsic should return what's available to the compiler but the user should also mark the function that uses the result, as well as any indirect user of the result (e.g some caller and callee of this function that uses the detection results) as "create additional clone for xyz features".

> would it make sense to add this as an LLVM intrinsic?

Depending on which flavor you want. For the former flavor (one that returns current compilation feature set) it doesn't have a gcc extension correspondence but might be generally useful for other frontends because it has a well defined semantics. (though I haven't heard of any, not that I looked very hard). For the latter flavor (the one that returns runtime feature set), it's what clang will need if/when they implement `target_clones` attribute but their semantics will be different from ours even if we implement this flavor here, and will definitely use different lowering (the info will come from different places) so it wouldn't help the pass very much.

Reply:
> . The intrinsic should return what's available to the compiler but the user should also mark the function that uses the result, as well as any indirect user of the result (e.g some caller and callee of this function that uses the detection results) as "create additional clone for xyz features". 

This sounds similar to https://rust-lang.github.io/rfcs/2045-target-feature.html / https://doc.rust-lang.org/reference/attributes/codegen.html#the-target_feature-attribute

So the example that Jameson showed me of a current usage that is non-local is:

```
fma(x::Float32, y::Float32, z::Float32) = Core.Intrinsics.have_fma(Float32) ? fma_llvm(x,y,z) : fma_emulated(x,y,z)

@inline function two_mul(x::T, y::T) where T<: Union{Float16, Float32}
    if Core.Intrinsics.have_fma(T)
        xy = x*y
        return xy, fma(x, y, -xy)
    end
    xy = widen(x)*y
    Txy = T(xy)
    return Txy, T(xy-Txy)
end
```

If the function `fma` is baked into the sysimage and the function `two_mul` isn't we have a cornercase if the sysimage has been compiled with `-fma`. The runtime query would answer "Yes this processor has fast fma", and so using it to decide whether or not the copy of fma is fast is faulty.

Reply:
Right, this is not a hypothetical, it is based specifically on the usages we have now.

>  as in it'll blow up exponentially

There are not ever going to be exponential processors produced. We also can gate the runtime feature based on the sysimage (similar to vector length).

The relevant question here is only whether we gate our `have_fma` pseudo-feature only, or whether we guarantee that we will feature-gate access to `fma` itself also to be identical to the value of `have_fma`.

Reply:
Currently we fail to feature gate `have_fma` to have any particular relationship to `fma` (it is neither conservative nor optimistic), except in the current function being compiled. That situation is not good.

Reply:
> So the example that Jameson showed me of a current usage that is non-local is:

But returning different values here is also bad since fma still won't be available in `fma`. The issue is what target `fma` is compiled for, not what value the `has_fma` returns in either function. As long as `fma` julia function is compiles without `fma` feature, it won't work correctly in this usage pattern.

In this particular case it might still have the expected behavior only because the libm implementation of `fma` might be doing the dispatch. `fma_llvm` still won't use fma instructions. (and FWIW the openlibm version isn't...)

> There are not ever going to be exponential processors produced. We also can gate the runtime feature based on the sysimage (similar to vector length).

Without a well defined feature dependency there is exponentially many processors that needs to be supported. That exponential space can only be reduced if you know all the processors that exists. Even if it's possible to create a way to reduce things to a minimal set, it's still not feasible to ensure future compatibility for the processor feature set database used to generate such reduction map.

> The relevant question here is only whether we gate our have_fma pseudo-feature only, or whether we guarantee that we will feature-gate access to fma itself also to be identical to the value of have_fma.

No we should not gate anything like that. For now, it's only about gating the access to fma, but soon enough you'll want to gate access to all the features. That basically means that the runtime feature has to be identical to the sysimg feature and that's exactly what I was setting out to avoid when I write the dispatch logic.

As I said, the right fix to this is really for the user to clone the right functions (returning the runtime value doesn't actually work and gating the feature set is way too restrictive). That's in general a call for the user, but in this case and at least as a stop gap measure it would be fine to make that decision in the compiler for fma. We can have a local cloning and dispatch just for fma's (or the different flavors of it) for functions that uses `have_fma` or `fma` (at llvm level). The dispatch logic will basically need to write a flag to let the caller to pick the right function from the sysimg.